Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -2124,6 +2130,206 @@ - (void)didTapFailedOutgoingMessage:(TSOutgoingMessage *)message
+
+    if (itemRange.location == NSNotFound) {
+        OWSFail(@"%@ Couldn't find range of quoted reply.", self.logTag);
+        return;
+    }
+
+    NSInteger desiredWindowSize = MAX(0, 1 + (NSInteger)threadInteractionCount - (NSInteger)itemRange.location);
+    NSUInteger oldLoadWindowSize = [self.messageMappings numberOfItemsInGroup:self.thread.uniqueId];
+    NSInteger additionalItemsToLoad = MAX(0, desiredWindowSize - (NSInteger)oldLoadWindowSize);
+
+    NSInteger dstIndex = 0;
+    if (additionalItemsToLoad > 0) {
+        // Try to load more messages so that the quoted message
+        // is in the load window.
+        //
+        // This may fail if the quoted message is very old, in which
+        // case we'll load the max number of messages.
+        [self loadNMoreMessages:(NSUInteger)additionalItemsToLoad];
+        
+        // Scroll to the first item, which should be the quoted message,
+        // or if we couldn't load it, the oldest loadable message.
+        //
+        // We use `indexPathRowOfQuotedMessage` instead of doing
+        // some index math because:
+        //
+        // * Resetting the mapping may integrate other pending changes.
+        // * Dynamic interactions may change.
+        dstIndex = (NSInteger)[self indexPathRowOfQuotedMessage:quotedMessage];
+    } else {
+        // Scroll to the quoted message, which is already in the load window.
+        dstIndex = 1 + (NSInteger)oldLoadWindowSize - desiredWindowSize;
+    }
+
+    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:dstIndex inSection:0];
+    [self.collectionView scrollToItemAtIndexPath:indexPath
+                                atScrollPosition:UICollectionViewScrollPositionTop
+                                        animated:YES];
+
+    // TODO: Highlight the quoted message?
+}
+
+// This method makes a best effort to determine the best scroll position
+// after loading more interactions when trying to scroll to a quoted
+// message.
+//
+// It prefers to be simple than to correctly handle edge cases like multiple
+// interactions with the same timestamp.
+- (NSUInteger)indexPathRowOfQuotedMessage:(TSQuotedMessage *)quotedMessage
+{
+    OWSAssertIsOnMainThread();
+    OWSAssert(quotedMessage);
+    OWSAssert(quotedMessage.timestamp > 0);
+    OWSAssert(quotedMessage.authorId.length > 0);
+
+    NSUInteger result = 0;
+    for (NSUInteger row = 0; row < self.viewItems.count; row++) {
+        ConversationViewItem *viewItem = self.viewItems[row];
+        if (viewItem.interaction.timestamp > quotedMessage.timestamp) {
+            break;
+        }
+        result = row;
+    }
+    return result;
+}
+
+// If result range is valid, then threadInteractionCount will have a valid value too.
+- (NSRange)findRangeOfQuotedMessage:(TSQuotedMessage *)quotedMessage
+             threadInteractionCount:(NSUInteger *)threadInteractionCount
```

Comment:
see above, I'm not sure that we need this method.

Reply:
See above.